### ISSUE TITLE
Improve RCDevice state and error reporting in the UI

### ISSUE THREAD

**Issue by atsakiridis** [RestComm/restcomm-ios-sdk#331]
Email discussion with @deruelle :

Sadly we do have some issues in that area :(. For example, if you start from a connected state and change registrar for example to an invalid one, it will take I think 32 seconds (default timeout for SIP I believe) to notify the user of the error code and update the internal state. Meanwhile the user can make calls (which will fail off course). After that we sporadically enter a corrupt state. Also, the messages are not very descriptive for ios after all. We only notify whether there is connectivity or not irrespective of whether the problem is registration or actual connectivity (Android is more descriptive).

So I'm thinking to open an issue and come up with a more concrete plan of action on those issues (how to handle the big SIP timeout, come up with user friendly yet descriptive messages, take care of any holes in the code), that will be consistent between iOS and Android.

Impacts (I'm filling those as I go):
* Add specific error reporting for INVITE, MESSAGE and REGISTER on these errors: authentication, time out as well some more specific for INVITE and MESSAGE, like 'not found'. Right now we have very generic categories and user can't tell what is going on.
* Unify error reporting of signaling module with RCDevice module cause it tends to be confusing. All error codes should reside at RestCommClient.h (which they are currently) and even the signaling module should reference those, not define new. 
* Make errors more user friendly. Avoid using SIP terminology
* When user changes registration details by calling RCDevice.updateParams() and this results in an actual registration we should change the RCDevice state to 'offline' so that we get back an event when we get online again
* Maybe also add a spinning wheel next to the icon when we are waiting for a response in a REGISTER
* Change the SIP Timer B in sofia from 32 secs (i.e. T1x64) to 8 secs as 32 is too long even for very slow networks. That way a timeout error will be issued much faster and the user will be in a position to identify it better
* Add error reporting for ssc_oper_create*() functions in SIP module, cause currently nothing is returned to the user when for example the SIP URI is invalid
* When either changing registrar or transitioning from non registrar-less to registrar-less then only unregister with previous registrar if we were actually online previously.
* When changing from non registrar-less to registrar-less if non registrar-less was not registered successfully previously, we should get a notification and transition to online (behind the scenes the Client library should not try to unregister with Restcomm)
* When updating only password or identity in the settings, without changing the registrar, we should still update those settings. Up to now those settings are only affected together with a registration with causes all sorts of errors
* Update reachability status of RCDevice when out of the App but a reachability event fires while in thee background
* Notify the App with a connectivity update when Device.listen() fails due to not having connectivity
* Seems deviceDidInitializeSignaling() can be covered by deviceDidStartListeningForIncomingConnections() and make API cleaner by removing the former.
* Cleanup Olympus state handling
* Avoid sending connectivity updates when deviceDidStartListeningForIncomingConnections() fires
* When we are resuming the App by pulling down the Control Center after having disabled Wifi before the OS notifies the Restcomm Client library, the App isn't isn't notified and thinks it's still online. Need to fix that
* Deliver the RCDevice didStart/StopListeningForIncomingConnections events asynchronously (that is using a selector to run after 0 seconds), to avoid calling the delegate functions directly from RCDevice methods, which seems a bit unorthodox and might cause issues down the road.

Important note: 

With the fixes so far issues seem to be resolved, but I'm getting some notification from the Client library that need to be better handled by the App, and maybe improved in the library. However, it will be more effective to revisit those once backgrounding is in place (i.e. https://github.com/RestComm/restcomm-ios-sdk/issues/153) cause most of those fire mostly due to the fact that we don't support backgrounding and the signaling facilities are shut down whenever the App is interrupted

**Comment by atsakiridis** [comment]
Reopening cause we miss some updates when the app starts up and the provided password is wrong. Seems that RCDeviceConnectivityType is not enough for the App. The also need the RCDeviceState when didStopListeningForIncomingConnections fires.

**Comment by atsakiridis** [comment]
Closed with 726aebe69ed0db9538fdf21d1269458fb14e32f8 on private branch at this point

**Comment by atsakiridis** [comment]
Reopening cause I found some more issues in the state:
* After opening Olympus after long time of it being in the background a. it shows up orange (i.e. connected), b. no notification pops to show whether registration succeeded and c. If I got to Settings to trigger a registration I get no notification about the registration status. Finally, when I return I see the icon grey and loading. From that point on there's no way I can make a call
* I think that if restcomm goes of line for some reason and online again there's a chance that we lose state

**Comment by atsakiridis** [comment]
Android Olympus doesn't receive incoming calls from Restcomm

**Comment by atsakiridis** [comment]
Found an issue not sure if it's the one described in the previous comment. Steps to reproduce:

1. Open Olympus, disable wifi and wait for it to be reflected in the UI
2. Leave Olympus with Home button
3. Enable wifi and wait for it to be enabled
4. Open Olympus

At that point state is lost and we cannot recover

**Comment by atsakiridis** [comment]
Found another issue when changing from wifi to cellular and then back to wif. Steps to reproduce:

* Start inside app, starting with wifi enabled, 
* Disable wifi while cellular is enabled, 
* We go offline because cellular times out to register (because restcomm is behind lan), 
* We re-enabled wifi but we remain off line for some reason

Analysis showed the following states:
* Online with wifi
* Pull control center and disable wifi: unlisten() is called: unregister & shutdown sofia
* Reachability changes to cellular
* Sofia starts up and binds to cellular interface
* Registration times out
* Pull control center: unlisten() is called but sofia isn't shutdown
* Reachablity changes to wifi
* Sofia is attempted to start up but is already up so network facilities aren't refreshed and hence the issue

**Comment by atsakiridis** [comment]
Also some error texts for Text Messages where wrong need to fix

**Comment by atsakiridis** [comment]
Now got another issue after long period of inactivity. To reproduce:
* Open Olympus and go online using wifi
* Hit home and leave it until the app get suspended iOS. Not sure how long that is. Last time I reproduced I think I opened the App after 2-3 hours

At that point I get nothing in the logs when the App starts, while when I hit home again I see that unlinsten() is called

**Comment by atsakiridis** [comment]
Also, another issue when changing registrar we should notify the app that we transitioned to offline until the registration succeeds or fails

**Comment by atsakiridis** [comment]
Another issue is:
* While off line open olympus and change registrar (from local to staging for example)
* Leave olympus and turn one network
* Enter olympus and the new settings will fail

Problem is that upon re-opening olympus the old registrar is tried
